DESIGN.md

```diff
@@ -368,3 +368,232 @@ http://docs.oracle.com/javase/specs/jls/se7/html/jls-8.html#jls-8.1
+
+Every piece of data in a GraphQL response can be identified by a query path from the root. In the query below, examples of query paths would be `hero.name` or `hero.friends[2].name`.
+
+```graphql
+query HeroAndFriendsNames {
+  hero {
+    name {
+      friends {
+        name
+      }
+    }
+  }
+}
+```
+
+Of course, query paths also need to take field arguments into account:
+
+```graphql
+query HeroName {
+  hero(episode: JEDI) {
+    name
+  }
+}
+```
+
+Here, the query path is `hero(episode:JEDI).name`, and that is (potentially) different from `hero(episode:EMPIRE).name`.
+
+Using query paths as cache keys gives us a way to decompose every result tree into flat records without relying on any additional conventions. Of course, in many cases this is not enough to avoid data duplication completely.
+
+In the Star Wars schema for instance, `hero`, `hero(episode:JEDI)` and `hero(episode:NEWHOPE)` refer to the same object (R2-D2), but these records would be cached separately. This is why schemas often include an `id` or other identifier field for types.
+
+Relay [relies on `id` to be always present and to be globally unique (so between types)](https://facebook.github.io/relay/docs/graphql-object-identification.html#content). Apollo allows you to specify your own [`dataIdFromObject` function](http://dev.apollodata.com/react/cache-updates.html#dataIdFromObject).
+
+### Runtime knowledge of query structure
+
+A consequence of this model is that computing a cache key for normalized caching can't just rely on the response. Let's say you receive this result:
+
+```json
+{
+  "data": {
+    "r2": {
+      "name": "R2-D2"
+    },
+    "luke": {
+      "name": "Luke Skywalker"
+    }
+  }
+}
+```
+
+In order to know the cache key for `r2` or `luke`, we need to know what field they are an alias for. In this case, `r2` should be cached under `hero`, and `luke` under `hero(episode:EMPIRE`:
+
+```graphql
+query TwoHeroes {
+  r2: hero {
+    name
+  }
+  luke: hero(episode: EMPIRE) {
+    name
+  }
+}
+
+```
+
+So we need runtime knowledge of the original query structure in order to perform normalized caching. One option would be to keep a separate representation of the query structure, as JSON for example. But that means we'd have to read in and traverse that structure when we parse a response, and we already have a representation of the query structure in the form of the generated model code.
+
+### Interleaving parsing and caching
+
+We'll have to discuss how this translates to what we're trying to do in Apollo Android, but in Apollo iOS the solution we've converged on is to interleave parsing and caching.
+
+Every model is able to instantiate itself from a `GraphQLResultReader` that can be used to read individual field values from either a network response or the normalized cache. When reading from a network response, it is also responsible for keeping track of the normalized records.
```

Comment:
I don't know enough about the way serialization libraries work on Android, but it would be nice if we could still have the ability to parse models from the normalized cache (without materializing a full JSON result structure) and to keep track of normalized records while parsing a response.